### ISSUE TITLE
13.0.7 - minified react error #421 (does not happen in 13.0.6)

### ISSUE THREAD

**Issue by wadehammes** [vercel/next.js#44083]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

 Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:03:51 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.18.1
      npm: 8.19.2
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.0.6
      eslint-config-next: 13.0.6
      react: 18.2.0
      react-dom: 18.2.0

### To Reproduce

Have a site running 13.0.7 with SSR, dynamic routes, and transpiled modules.

### Describe the Bug

Recieve the below console error in both preview and local.

https://reactjs.org/docs/error-decoder.html/?invariant=421

<img width="1140" alt="Screenshot 2022-12-16 at 9 53 20 AM" src="https://user-images.githubusercontent.com/222170/208125397-f3ceec24-85b6-4b03-9df8-c750d6d3e861.png">

### Expected Behavior

No suspense boundary error as in 13.0.6

### Which browser are you using? (if relevant)

Firefox

### How are you deploying your application? (if relevant)

Vercel

**Comment by gusfune** [comment]
Can confirm the same issue here. The full error is:
> Error: This Suspense boundary received an update before it finished hydrating. This caused the boundary to switch to client rendering. The usual way to fix this is to wrap the original update in startTransition.

Happens even in blank components all random in the latest NextJS version.

**Comment by wegeberg** [comment]
Same happens here in multiple apps when running in production.
Debian 11 - Node 16.18.1  - x86_64
Downgrading to Next 13.0.6 removes the error in all apps.

**Comment by kachkaev** [comment]
Also seeing this in 13.0.7.

### `yarn dev`

`next/dynamic` + `ssr: true` (default)

```
react-dom.development.js?ac89:20702 Uncaught Error: This Suspense boundary received an update before it finished hydrating. This caused the boundary to switch to client rendering. The usual way to fix this is to wrap the original update in startTransition.
    at updateDehydratedSuspenseComponent (react-dom.development.js?ac89:20702:1)

Error: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.
```

### `yarn build && yarn start`

`next/dynamic` + `ssr: true` (default)

```
Uncaught Error: Minified React error #421; visit https://reactjs.org/docs/error-decoder.html?invariant=421 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
```

`next/dynamic` + `ssr: false` 

```
Uncaught Error: Minified React error #419; visit https://reactjs.org/docs/error-decoder.html?invariant=419 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
```

Reverting to 13.0.6 helps.

**Comment by huozhi** [comment]
The link in the PR description is not accessible with password protected. Could someone provide a mini reproduction repo for this?

**Comment by wadehammes** [comment]
@huozhi the error here isn't the exact error but I believe it relates: https://codesandbox.io/s/awesome-hermann-gu8kmj?file=/pages/index.tsx

I think this has to do with the combination of `getStaticProps` and using `next/dynamic` as reported by @kachkaev (and potentially the use of `styled-components`)

**Comment by wadehammes** [comment]
Ive started to work on a reproduction here -> https://codesandbox.io/s/awesome-hermann-gu8kmj?file=/pages/index.tsx

It appears to be caused (for my use case anyways) trying to dynamically load a styled-component with next/dynamice.

**Comment by huozhi** [comment]
> Ive started to work on a reproduction here -> https://codesandbox.io/s/awesome-hermann-gu8kmj?file=/pages/index.tsx
I didn't see warnings in the codesandbox, also when I cloned it and tried with next dev/build/start, they also don't have warnings either.

**Comment by wadehammes** [comment]
You are correct. This is not the same error then. Do you want access to my company repo to look into the error reported?

**Comment by wadehammes** [comment]
If someone else here is able to create a reproduction that would be great.

**Comment by gusfune** [comment]
Ok, I have managed to narrow down the culprit. Unfortunately I couldn't create a reproduction for this, but I can confirm the problem is specifically with `next/dynamic`.

The error occurs on components with dynamic when no option it set, or when the option contains `suspense: true`.
Changing the `dynamic` import to `ssr: false` solves the issue.

**Comment by yarigo** [comment]
I was able to localize the error in my project and reproduce it (with redux): https://codesandbox.io/p/sandbox/next-13-0-7-swkcsy.
In version 13.0.6 everything works.
In version 13.0.7 if you set `ssr: false` for component **A** or import it directly (without dynamic), everything works.

**Comment by wadehammes** [comment]
@huozhi see above for a reproduction

**Comment by huozhi** [comment]
This error is removed in https://github.com/facebook/react/pull/25692, I tried `18.3.0-next-81d4ee9ca-20221223` of react it doesn't show the warning in codesandbox. Seems 18.3 is going to be released soon

**Comment by conquera99** [comment]
so its because of react?

I tried simple app, but the error is shown too (using `appDir`).

package:
nextjs: `13.1.1`
react: `18.2.0`

https://github.com/conquera99/website

**Comment by huozhi** [comment]
The error mentioned in https://github.com/vercel/next.js/issues/44083#issuecomment-1355042765 is removed in react. If it's still happening might be sth else

**Comment by huozhi** [comment]
@conquera99 your issue is from `react-social-icon`, in dev mode you can see there's a hydration mistmatch issue, and it asked you to not nested `<a>` inside `<a>`. You need to remove the wrapping anchor to `react-social-icons`, then pass the props on anchor to `ScoialLink` itself.

**Comment by conquera99** [comment]
ahh my bad, thank you @huozhi

**Comment by phuctm97** [comment]
I'm not sure if it is related. `dynamic` imported components inside a `Suspense` component don't render its fallback, it just doesn't render anything at all when being suspended. The issue happens in 13.0.7 as well as latest version, downgrading to 13.0.6 get rid of the issue.

**Comment by oceandrama** [comment]
@huozhi As mentioned here https://github.com/facebook/react/pull/25692, underlying issue is still there
Even error is not visible, it leads to full rerendering application on the client side
I have the same error with `next/dynamic` in Next.js v13.1.1, but not in v 13.0.6

**Comment by wadehammes** [comment]
Nice find @oceandrama

**Comment by tatethurston** [comment]
Encountering this on Next.js v13.1.1.

Adding `{ ssr: false }` to all `next/dynamic` call sites temporarily works around the issue -- if only importing dynamic components on the client is acceptable for your use case.

Example:

```js
import dynamic from 'next/dynamic';

const LazyComponent = dynamic(
  () => import('./Component').then((mod) => mod.Component),
  { ssr: false },
);
```

**Comment by tatethurston** [comment]
Updating to React 18.3 appears to fix the issue.

**Comment by phuctm97** [comment]
> Updating to React 18.3 appears to fix the issue.

How to get access to React 18.3? I checked on React repository but the latest available version is 18.2.

**Comment by tatethurston** [comment]
> > Updating to React 18.3 appears to fix the issue.
> 
> How to get access to React 18.3? I checked on React repository but the latest available version is 18.2.

react@next

**Comment by wadehammes** [comment]
> Updating to React 18.3 appears to fix the issue.

This only hides the error, but the issue is still there underlying. See first line of the PR for the change in React 18.3 -> https://github.com/facebook/react/pull/25692

**Comment by klarstrup** [comment]
This is still the case as of `next@13.1.2-canary.4` with `react@18.2`.

`next@13.1.1`(and indeed `next@13.1.2-canary.4`) with `react@18.3.0-next-0f4a83596-20230110` quenches the error.

**Comment by wadehammes** [comment]
> This is still the case as of `next@13.1.2-canary.4` with `react@18.2`.
> 
> `next@13.1.1`(and indeed `next@13.1.2-canary.4`) with `react@18.3.0-next-0f4a83596-20230110` quenches the error.

@klarstrup the error is removed, but the problem still exists. I'm personally not planning to upgrade to 13.1.1 or beyond even if 18.3 releases until we get word from the Next team they plan to fix the underlying issue. @balazsorban44 @huozhi

**Comment by klarstrup** [comment]
> @klarstrup the error is removed, but the problem still exists. I'm personally not planning to upgrade to 13.1.1 or beyond even if 18.3 releases until we get word from the Next team they plan to fix the underlying issue. @balazsorban44 @huozhi

Yes I specifically say it "quenches the error" to emphasize this ☺️

**Comment by huozhi** [comment]
@yarigo I tried your reproduction with `react@next` and `react-dom@next` it works since the error is removed in the react next channel. But I found disabling react devtools works for that. I filed an issue for it on react side https://github.com/facebook/react/issues/25994

**Comment by yarigo** [comment]
@huozhi Yes, everything works fine with these versions:
```
"next": "^13.1.2"
"react": "^18.3.0-next-555ece0cd-20230112"
"react-dom": "^18.3.0-next-555ece0cd-20230112"
```

**Comment by anjan-kp** [comment]
This issue is related to dynamic import in nextjs. 
Looks like there is a version compatibility issue between next@latest and react@latest
Working fine in these 2 version combinations of react and nextjs.
next:13.0.6
react:18.2.0
react-dom:18.2.0

And

next@latest
react@next
react-dom@next

**Comment by wadehammes** [comment]
Im guessing there will not be a fix for this. Still happening in 13.1.3.

**Comment by huozhi** [comment]
@wadehammes can you provide your mini reproduction? since the problems for above ones are resolved. need a repro for seeing how the prod error is triggered

**Comment by phuctm97** [comment]
> People are avoiding upgrading their Next.js version or downgrading altogether to get around this issue, which is ironic because the PR was supposed to make upgrading/migration to the _app_ folder easier. It may be appropriate to revisit the changes in this PR.

I can't agree more. This is a big breaking change for all of my Next.js apps. Can't upgrade at all since 13.0.7.

**Comment by wescopeland** [comment]
> We are trying out best here.

I appreciate that, nice new website btw.

> it is definitely happening to some folks (and we want to continue looking into this with reproductions) but not a majority.

My perception is this is not happening on simple sites, but more complex webapps, which the majority of users are not building. This is also likely why it is harder to get a consistent repro and/or why each case has had subtle differences.

@phuctm97 Do you have a repro you can share? I am only able to repro this in more-complex apps that are unfortunately locked down.

**Comment by phuctm97** [comment]
Thanks for proceeding on this @leerob & @wescopeland. I have reproduced the issue here: https://github.com/phuctm97/repro-next-44083.

The repository is created using `yarn create next-app --ts`, all with latest version. There are comments in the repo explaining the issues.

**Comment by huozhi** [comment]
@phuctm97 I can see the issue in the reproduction, it's because the nested `next/dynamic` is not preloaded, we were aware of that issue and exploring solution is ongoing. But seems I couldn't repro the minified react error or any error in dev mode, I wonder it's caused by sth else

**Comment by phuctm97** [comment]
Thanks for updating @huozhi. Any suggested workaround or issue to track that issue. It is a big breaking blocker for me at the moment.

**Comment by huozhi** [comment]
@phuctm97 Since we introduced a layer of `Suspense` in dynamic, you don't have to wrap extra Suspense boundary to it. And you can put the `Fallback` component in `loading` option
<img width="589" alt="image" src="https://user-images.githubusercontent.com/4800338/216632673-deb827b5-3a3b-489a-951a-8b5fe0534cc3.png">

**Comment by phuctm97** [comment]
Thanks @huozhi! That works for simple fallback but if `<Example />` takes some props and `<Fallback />` needs some of these props to render a higher fidelity fallback UI, a React context will need to be used in this pattern. Using `<Suspense />` is more natural & flexible. Just a feedback, using context should work for now!

**Comment by huozhi** [comment]
> <Fallback /> needs some of these props to render a higher fidelity fallback UI, a React context will need to be used for each instance of this pattern. Using <Suspense /> is more natural & flexible
@phuctm97 I would recommend to use React.lazy and Suspense together, but it also requires some change like converting to static html rendering for `pages/`, which we have planned to put it in later releases.

Meanwhile, #45565 should resolve the problem since it reverts to the original implementation of dynamic for pages

**Comment by huozhi** [comment]
Got the fix out in v13.1.7-canary.5, can you try that version to see if it works for you? Thanks 🙏

**Comment by KittySparkles** [comment]
I can confirm this appears to be fixed with v13.1.7-canary.5. ✨

**Comment by phuctm97** [comment]
I can confirm the issue appears to be fixed with v13.1.7-canary.5 too!

**Comment by icekhab** [comment]
It works without `Suspense boundary error` with v13.1.7-canary.5, thanks!
